core/ray.h

```diff
@@ -41,24 +37,51 @@ namespace core
     struct RayHitData
     {
         RayHitData() = default;
-        RayHitData(float t, const Vec3 &norm, const Ray &ray, float freq) : time(t), normal(norm), frequency(freq), collisionPoint(ray.at(t)), direction(ray.getDirection()), energy(ray.getEnergy()), phase(ray.phaseAt(freq, t))
-        {
-            origin = ray.getOrigin();
-        }
+        RayHitData(float t, const Vec3 &norm, const Ray &ray, float freq)
+         : time_(t), ray_(ray), normal_(norm), frequency_(freq) {};
         ~RayHitData() = default;
         RayHitData(const RayHitData &) = default;
 
-        // Operators? No kidding!
-        // These comments are just visual clutter. I would remove them. It is more important to e.g
-        // describe what this does if the impl. is somehow not really self explanatory (e.g skips some params).
-        // OPERATORS
         bool operator==(const RayHitData &other) const;
         friend std::ostream &operator<<(std::ostream &os, const RayHitData &rayData);
 
-        // VARIABLES
-        // Why do you store the members of the Ray separately?
-        Vec3 collisionPoint, direction, normal, origin;
-        float time, energy, phase, frequency;
+        Vec3 normal() const
```

Comment:
I thought this will be better, because this way I will calculate only things that I need at a time.

Reply:
That sounds good e.g with collisionPoint(), but what about normal, time and frequency? It is exactly as before.

Also for things like Direction, it may be better to just allow to return the ray that hit right? insted of unpacking it.

You do not have to use the always the samething for everything, use the technique that works the best for the case.

Reply:
done